### PR TITLE
[SPARK-13048][ML][MLLIB] keepLastCheckpoint option for LDA EM optimizer

### PR DESCRIPTION
**Pull request by jkbradley**
## What changes were proposed in this pull request?

The EMLDAOptimizer should generally not delete its last checkpoint since that can cause failures when DistributedLDAModel methods are called (if any partitions need to be recovered from the checkpoint).

This PR adds a "deleteLastCheckpoint" option which defaults to false.  This is a change in behavior from Spark 1.6, in that the last checkpoint will not be removed by default.

This involves adding the deleteLastCheckpoint option to both spark.ml and spark.mllib, and modifying PeriodicCheckpointer to support the option.

This also:
- Makes MLlibTestSparkContext extend TempDirectory and set the checkpointDir to tempDir
- Updates LibSVMRelationSuite because of a name conflict with "tempDir" (and fixes a bug where it failed to delete a temp directory)
- Adds a MIMA exclude for DistributedLDAModel constructor, which is already `private[clustering]`
## How was this patch tested?

Added 2 new unit tests to spark.ml LDASuite, which calls into spark.mllib.

### CODE REVIEW

**Comment by jkbradley** [comment]
Hold on...I'm also switching from deleteLastCheckpoint to keepLastCheckpoint

**Comment by jkbradley** [comment]
OK should be good to go now

**Comment by jkbradley** [comment]
@hhbyyh @holdenk Would you mind taking a look please?  Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala [hunk 1]
```diff
@@ -619,6 +651,31 @@ class DistributedLDAModel private[ml] (
...
+  /**
+   * If using checkpointing and [[LDA.keepLastCheckpoint]] is set to true, then there may be
+   * saved checkpoint files.  This method is provided so that users can manage those files.
+   * Note that removing the checkpoints can cause failures if a partition is lost and is needed
+   * by certain [[DistributedLDAModel]] methods.
+   *
+   * @return  Checkpoint files from training
+   */
+  @Since("2.0.0")
+  def getCheckpointFiles: Array[String] = _checkpointFiles
```
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala]: Do we really want to expose this? It seems somewhat closely tied to the implementation - maybe we should mark this as a developer API?

File: mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala [hunk 2]
```diff
@@ -258,7 +265,30 @@ private[clustering] trait LDAParams extends Params with HasFeaturesCol with HasM
   def getOptimizeDocConcentration: Boolean = $(optimizeDocConcentration)
 
   /**
+   * For EM optimizer, if using checkpointing, this indicates whether to keep the last
+   * checkpoint. If false, then the checkpoint will be deleted. Deleting the checkpoint can
+   * cause failures if a data partition is lost, so set this bit with care.
+   *
+   * See [[DistributedLDAModel.getCheckpointFiles]] for getting remaining checkpoints and
+   * [[DistributedLDAModel.deleteCheckpointFiles]] for removing remaining checkpoints.
+   *
```
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala]: Maybe add a note that by default reference tracking will eventually clean up the checkpoint regardless so people don't set this to false unnecessarily.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala [hunk 3]
```diff
@@ -758,6 +816,10 @@ class LDA @Since("1.6.0") (
   @Since("1.6.0")
   def setOptimizeDocConcentration(value: Boolean): this.type = set(optimizeDocConcentration, value)
 
+  /** @group expertSetParam */
+  @Since("2.0.0")
+  def setKeepLastCheckpoint(value: Boolean): this.type = set(keepLastCheckpoint, value)
+
```
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala]: Do we need to keep the setter in model?

File: mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala (hunk 1 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala]: I'd like to give users a way to clean up manually, but I'll mark it as DeveloperApi.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala (hunk 2 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala]: Sounds good.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala (hunk 3 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala]: I don't think so.  Once there is a model, the decision about keeping the last checkpoint has already been made.  Users can manage the checkpoint via the deletion method though.

**Comment by jkbradley** [comment]
Updated!

**Comment by holdenk** [comment]
This change looks good to me pending tests/MiMa & possibly user verification - we should also either add a follow up JIRA for the Python API or add it in this PR.

**Comment by jkbradley** [comment]
It looks like the MiMa tests were spurious (Spark core)
I'll make a follow-up JIRA for Python

**Comment by jkbradley** [comment]
Fixed merge conflicts in MimaExcludes.scala by rebasing.

**Comment by jkbradley** [comment]
Merging with master
@holdenk @hhbyyh Thanks for taking a look!

**Pull request closed without merging**